### DISCUSSION TITLE
Multiple inheritance with slotted parent

### DISCUSSION THREAD

**Discussion by aminalaee** [PyO3/pyo3#3547] [Questions] [answered]
My knowledge of Python data model is pretty limited, but I was expecting the following would work:
https://docs.python.org/3/reference/datamodel.html#slots

```rust
#[pyclass(subclass)]
#[derive(Clone, Debug)]
struct A {}

#[pymodule]
fn my_module(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<A>()?;
    Ok(())
}
```

```py
from my_module import A

class B:
    __slots__ = ('example',)


class C(A, B): ...
```

This fails with:

```python-traceback
>       class C(A, B):
E       TypeError: multiple bases have instance lay-out conflict
```

But the rust class `A` doesn't define the slots, or even if I set it to empty, this still fails but I expected multiple inheritance should work.
If `A` was defined in Python, the same thing would work.

Can anyone point me to the right direction of what is happening?

**Comment by davidhewitt** [comment, marked as answer]
So the cause here is that PyO3 still adds some extra space to `struct A`'s Python layout to do our refcell tracking. You can turn this off with `#[pyclass(frozen)]`. (We've discussed making frozen the default in the past.)

Testing locally, I indeed find this works:

```rust
#[pyclass(subclass, frozen)] // <--- frozen added here
#[derive(Clone, Debug)]
struct A {}

#[pymodule]
fn my_module(_py: Python, m: &PyModule) -> PyResult<()> {
    m.add_class::<A>()?;
    Ok(())
}
```

Note that if you make `A` non-empty then again you'll have a conflict. That's just a limitation of Python multiple inheritance, there's not much that can be done about that.

**Reply by aminalaee** [reply]
Thank you for the answer.
